content/microservices/kafka.md

```diff
@@ -0,0 +1,283 @@
+
+To prevent the `ClientKafka` consumers from losing response messages a custom partitioner is utilized.  This custom partitioner assigns partitions to a collection of consumers sorted by high-resolution timestamps (`process.hrtime()`) that are set on application launch.
+
+#### Incoming
+
+Nest receives incoming Kafka messages as an object with `key`, `value`, and `headers` properties that have values of the `Buffer` type. Nest then parses these values by transforming the buffers into strings. If the string is "object like", Nest attempts to parse the string as `JSON`. The parsed object is then passed to its associated handler with the following properties.
+
+```typescript
+interface IncomingMessage {
+  topic: string;
+  partition: number;
+  timestamp: string;
+  size: number;
+  attributes: number;
+  offset: string;
+  key: any;
+  value: any;
+  headers: Record<string, any>;
+}
+```
+
+#### Outgoing
+
+Nest sends outgoing Kafka messages after a serialization process when publishing events or sending messages. This occurs on arguments passed to the client `emit()` and `send()` methods or on values returned from a `@MessagePattern` method. This serialization "stringifies" objects that are not strings or buffers by using `JSON.stringify()` or the `toString()` prototype method.
+
+```typescript
+@@filename(hero.controller)
+@Controller()
+export class HeroService {
+  @MessagePattern('hero.kill.dragon')
+  killDragon(message: any): any {
+    const dragonId = message.value.dragonId;
+    const items = [
+      { id: 1, name: 'Mythical Sword' },
+      { id: 2, name: 'Key to Dungeon' },
+    ];
+
+    return items;
+  }
+}
+```
+
+Outgoing messages can also be keyed by passing an object with the `key` and `value` properties. Keying messages is important for meeting the [co-partitioning requirement](https://docs.confluent.io/current/ksql/docs/developer-guide/partition-data.html#co-partitioning-requirements).
+
+```typescript
+@@filename(hero.controller)
+@Controller()
+export class HeroService {
+  @MessagePattern('hero.kill.dragon')
+  killDragon(message: any): any {
+    const realm = 'Nest';
+    const heroId = message.value.heroId;
+    const dragonId = message.value.dragonId;
+
+    const items = [
+      { id: 1, name: 'Mythical Sword' },
+      { id: 2, name: 'Key to Dungeon' },
+    ];
+
+    return {
+      headers: {
+        realm
+      },
+      key: heroId,
+      value: items
+    }
+  }
```

Comment:
I saw an example of outgoing messages on the tests that use the  `client` `emit()` and `send()`. Is this another way to send messages. Which one is preferred?